File: frontend/src/lib/local/engine.ts
/**
 * Running a program in the browser with jq-wasm (jq 1.8.2), for the static build. It answers
 * with the same `RunResult` the server does: the same stand-in `$ENV`, the same located
 * compile errors, the same formatting (jq's own), and the equivalent command line.
 */

import { equivalentCommand, outputFlags, readingFlags, readsInput } from '@/lib/local/command'
import { compileErrors, readStreams } from '@/lib/local/diagnostics'
import { isGeoJson } from '@/lib/local/geojson'
import type { WorkerReply, WorkerRequest } from '@/lib/local/jq.worker'
import type { JsonValue, RunOptions, RunResult } from '@/lib/types'

/** The environment programs see as `$ENV`, as on the server. */
export const SANDBOX_ENV = {
    HOME: '/home/learner',
    USER: 'learner',
    SHELL: '/bin/bash',
    PAGER: 'less',
    LANG: 'C.UTF-8',
    TZ: 'UTC',
}

const SANDBOX_PREFIX = `${JSON.stringify(SANDBOX_ENV)} as $ENV | def env: $ENV; (`
const SANDBOX_SUFFIX = '\n)'

/** What `--seq` prints before every output. */
const RECORD_SEPARATOR = String.fromCharCode(0x1e)

const TIMEOUT_MS = 4000
const MAX_OUTPUTS = 10_000

let worker: Worker | null = null
let sequence = 0
const waiting = new Map<number, (reply: WorkerReply) => void>()

function startWorker(): Worker {
    const created = new Worker(new URL('./jq.worker.ts', import.meta.url), { type: 'module' })
    created.onmessage = (event: MessageEvent<WorkerReply>) => {
        waiting.get(event.data.id)?.(event.data)
        waiting.delete(event.data.id)
    }
    return created
}

class Timeout extends Error {}

/** Run jq once in the worker, terminating it when the time limit is reached. */
function exchange(input: string, program: string, flags: string[]): Promise<WorkerReply> {
    worker ??= startWorker()
    const id = ++sequence
    const request: WorkerRequest = { id, input, program, flags }
    return new Promise((resolve, reject) => {
        const timer = setTimeout(() => {
            waiting.delete(id)
            worker?.terminate()
            worker = null
            reject(new Timeout())
        }, TIMEOUT_MS)
        waiting.set(id, (reply) => {
            clearTimeout(timer)
            resolve(reply)
        })
        worker?.postMessage(request)
    })
}

/** Every output value from compact stdout; `--seq` puts a record separator before each one. */
export function parseLines(stdout: string): { outputs: JsonValue[]; truncated: boolean } {
    const outputs: JsonValue[] = []
    for (const raw of stdout.split('\n')) {
        const line = raw.startsWith(RECORD_SEPARATOR) ? raw.slice(1) : raw
        if (line === '') continue
        if (outputs.length >= MAX_OUTPUTS) return { outputs, truncated: true }
        outputs.push(JSON.parse(line) as JsonValue)
    }
    return { outputs, truncated: false }
}

/** Run a program over an input text in the browser. */
export async function runLocally(program: string, input: string, options: RunOptions): Promise<RunResult> {
    const started = performance.now()
    const command = equivalentCommand(program, options, readsInput(program, options))
    const finish = (partial: Partial<RunResult>): RunResult => ({
        ok: (partial.errors ?? []).length === 0,
        engine: 'cli',
        outputs: [],
        text: '',
        errors: [],
        messages: [],
        truncated: false,
        is_geojson: false,
        command,
        exit_code: null,
        ...partial,
        duration_ms: Math.round((performance.now() - started) * 100) / 100,
    })
    if (Object.keys(options.modules).length > 0) {
        return finish({
            errors: [
                {
                    kind: 'unavailable',
                    message: 'modules (-L) need files on disk; run the playground locally to use them',
                    line: null,
                    column: null,
                    end_column: null,
                },
            ],
        })
    }
    const reading = readingFlags(options)
    try {
        // The browser has no files: a slurp file is bound the way jq binds it, as the array of its
        // values (jq itself reads them, so number literals keep their form), and a raw file as text.
        const variables: string[] = []
        for (const [name, value] of Object.entries(options.args)) variables.push('--arg', name, value)
        for (const [name, value] of Object.entries(options.argjson))
            variables.push('--argjson', name, JSON.stringify(value))
        for (const [name, content] of Object.entries(options.slurpfile)) {
            const slurped = await exchange(content, '.', ['-c', '-s'])
            if (slurped.exitCode !== 0) {
                return finish({
                    errors: [
                        {
                            kind: 'input',
                            message: `--slurpfile ${name}: ${slurped.stderr.replace(/^jq: /, '')}`,
                            line: null,
                            column: null,
                            end_column: null,
                        },
                    ],
                })
            }
            variables.push('--argjson', name, slurped.stdout.trim())
        }
        for (const [name, content] of Object.entries(options.rawfile)) variables.push('--arg', name, content)
        const bound = variables
        // Positional arguments reach the program through $ARGS, which jq builds from every variable.
        let argumentsPrefix = ''
        if (options.positional.length > 0) {
            const named = await exchange('null', '$ARGS.named', ['-n', '-c', ...bound])
            const positional = options.positional_json
                ? `[${options.positional.join(',')}]`
                : JSON.stringify(options.positional)
            argumentsPrefix = `{"positional": ${positional}, "named": ${named.stdout.trim()}} as $ARGS | `
        }
        const wrapped = SANDBOX_PREFIX + argumentsPrefix + program + SANDBOX_SUFFIX
        // One run for the values (compact JSON, one per line), one for the text as jq prints it.
        const values = await exchange(input, wrapped, ['-c', ...reading, ...bound])
        if (values.exitCode === 3) {
            // Compile errors are located against the author's program, not the sandbox wrapper.
            const bare = await exchange('null', program, [
                '-n',
                ...reading.filter((flag) => flag !== '-n'),
                ...bound,
            ])
            return finish({ errors: compileErrors(bare.stderr || values.stderr) })
        }
        const { outputs, truncated } = parseLines(values.stdout)
        const { errors, messages } = readStreams(values.stderr, values.exitCode)
        const printing = [
            ...readingFlags(options).filter((flag) => flag !== '--seq'),
            ...outputFlags(options),
            ...bound,
        ]
        const printed = await exchange(input, wrapped, printing)
        return finish({
            outputs,
            truncated,
            errors,
            messages,
            text: printed.stdout,
            exit_code: printed.exitCode,
            is_geojson: outputs.length === 1 && errors.length === 0 && isGeoJson(outputs[0]),
        })
    } catch (error) {
        if (error instanceof Timeout) {
            return finish({
                errors: [
                    {
                        kind: 'timeout',
                        message: `the program ran longer than ${String(TIMEOUT_MS / 1000)}s and was stopped`,
                        line: null,
                        column: null,
                        end_column: null,
                    },
                ],
            })
        }
        throw error
    }
}

File: src/playground_jq/jq/engine.py
"""Running one jq program: pick the engine, enforce the limits, format the result."""

import re
import time
from typing import Any, Literal, cast

from pydantic import JsonValue

from playground_jq.config import Settings
from playground_jq.jq.binary import locate, run_binary
from playground_jq.jq.diagnostics import compile_errors, plain_error
from playground_jq.jq.formatting import equivalent_command, format_outputs
from playground_jq.jq.models import SANDBOX_ENV, JqError, RunOptions, RunResult
from playground_jq.jq.runner import POOL, RunnerPool, RunnerStopped
from playground_jq.sources.geojson import is_geojson

#: Builtins only the command line gives their full meaning: the library evaluates every
#: input value on its own and has nowhere to send `debug`/`stderr`.
CLI_ONLY = re.compile(r"(?<![\w$.])(input|inputs|input_filename|input_line_number|debug|stderr|halt|halt_error)\b")


#: An integer literal longer than a double holds exactly; jq 1.8 keeps it, jq.py rounds it.
BIG_INTEGER = re.compile(r"(?<![\d.])\d{16,}(?![\d.])")


def needs_cli(program: str, options: RunOptions, input_text: str = "") -> bool:
    """Whether this program, its flags or its input need the command-line binary.

    `--seq` changes how input is read as well as how output is written, so it runs where jq
    implements both; large integer literals keep their precision only in jq itself.
    """
    if options.stream or options.seq or CLI_ONLY.search(program):
        return True
    return bool(BIG_INTEGER.search(program) or BIG_INTEGER.search(input_text))


def reads_input(program: str, options: RunOptions) -> bool:
    """Whether the run reads any input text at all."""
    return not options.null_input or bool(re.search(r"(?<![\w$.])inputs?\b", program))


async def run_program(
    program: str,
    input_text: str,
    options: RunOptions,
    settings: Settings,
    *,
    pool: RunnerPool = POOL,
) -> RunResult:
    """Run a program over an input text with the given flags."""
    started = time.perf_counter()
    engine: Literal["library", "cli"] = "library"
    if options.engine == "cli" or (options.engine == "auto" and needs_cli(program, options, input_text)):
        engine = "cli"
    command = equivalent_command(program, options, reads_input=reads_input(program, options))

    def finish(
        outputs: list[JsonValue], errors: list[JqError], *, truncated: bool = False, messages: list[str] | None = None
    ) -> RunResult:
        text = format_outputs(outputs, options)
        if len(text.encode()) > settings.jq_max_output_bytes:
            text = text.encode()[: settings.jq_max_output_bytes].decode(errors="ignore")
            truncated = True
        return RunResult(
            ok=not errors,
            engine=engine,
            outputs=outputs,
            text=text,
            errors=errors,
            messages=messages or [],
            truncated=truncated,
            duration_ms=round((time.perf_counter() - started) * 1000, 2),
            is_geojson=len(outputs) == 1 and not errors and is_geojson(outputs[0]),
            command=command,
        )

    if len(input_text.encode()) > settings.jq_max_input_bytes:
        limit = f"the input is larger than the {settings.jq_max_input_bytes} byte limit"
        return finish([], [JqError(kind="limit", message=limit)])
    if engine == "cli":
        executable = locate(settings.jq_binary)
        if executable is None:
            missing = f"this program needs the jq command line ({settings.jq_binary}), which is not installed"
            return finish([], [JqError(kind="unavailable", message=missing)])
        try:
            outcome = await run_binary(
                executable,
                program,
                input_text,
                options,
                timeout=settings.jq_timeout_seconds,
                max_outputs=settings.jq_max_outputs,
            )
        except TimeoutError:
            return finish([], [_timeout(settings)])
        return finish(outcome.outputs, outcome.errors, truncated=outcome.truncated, messages=outcome.messages)
    request: dict[str, Any] = {
        "program": program,
        "variables": {**options.args, **options.argjson},
        "input": input_text,
        "slurp": options.slurp,
        "null_input": options.null_input,
        "raw_input": options.raw_input,
        "max_outputs": settings.jq_max_outputs,
        "env": SANDBOX_ENV,
    }
    try:
        reply = await pool.run(request, settings.jq_timeout_seconds)
    except TimeoutError:
        return finish([], [_timeout(settings)])
    except RunnerStopped as error:
        return finish([], [JqError(kind="runtime", message=str(error))])
    outputs = cast("list[JsonValue]", reply.get("outputs", []))
    errors: list[JqError] = []
    if "error" in reply:
        message = cast("str", reply["error"])
        errors = compile_errors(message) if reply.get("phase") == "compile" else [plain_error("runtime", message)]
    for failure in cast("list[dict[str, str]]", reply.get("errors", [])):
        errors.append(plain_error("input" if failure["phase"] == "input" else "runtime", failure["error"]))
    return finish(outputs, errors, truncated=bool(reply.get("truncated")))


def _timeout(settings: Settings) -> JqError:
    """The error a run that ran out of time is answered with."""
    return JqError(
        kind="timeout", message=f"the program ran longer than {settings.jq_timeout_seconds:g}s and was stopped"
    )
